File: src/api-client.js
import axios from 'axios';
import dotenv from 'dotenv';
import https from 'https';
import { readFile } from 'fs/promises';

// Load environment variables
dotenv.config();

class VeniceApiClient {
  constructor() {
    this.apiKey = process.env.VENICE_API_KEY;
    if (!this.apiKey) {
      console.error('Error: VENICE_API_KEY not found in environment variables');
      throw new Error('VENICE_API_KEY is required. Set it in your .env file');
    }
    this.baseUrl = 'https://api.venice.ai/api/v1';
    this.model = process.env.MEDIUM_VENICE_MODEL || 'deepseek-r1-llama-70b';
  }

  /**
   * List available models
   * @returns {Promise<Array>} List of available models
   */
  async listModels() {
    try {
      const response = await axios({
        method: 'get',
        url: `${this.baseUrl}/models`,
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json'
        }
      });
      
      return response.data;
    } catch (error) {
      this._handleError(error);
    }
  }

  /**
   * Send a prompt to the Venice API
   * @param {Array} messages Array of message objects with role and content
   * @param {Object} options Additional options for the API call
   * @returns {Promise<Object>} API response
   */
  async sendPrompt(messages, options = {}) {
    try {
      const defaultOptions = {
        temperature: 0.7,
        max_tokens: 1000,
        stream: false,
      };

      const requestOptions = { ...defaultOptions, ...options };
      
      const response = await axios({
        method: 'post',
        url: `${this.baseUrl}/chat/completions`,
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json'
        },
        data: {
          model: this.model,
          messages,
          ...requestOptions
        }
      });
      
      return response.data;
    } catch (error) {
      this._handleError(error);
    }
  }

  /**
   * Generate nutrition-focused response
   * @param {string} userInput User's question or request
   * @returns {Promise<string>} LLM response
   */
  async getNutritionGuidance(userInput) {
    const systemPrompt = `You are a nutrition expert specializing in longevity-focused meal planning. 
    Provide evidence-based guidance on foods and dietary patterns that promote health and longevity. 
    Emphasize whole, nutrient-dense foods like those found in Blue Zones and the Mediterranean diet.
    When recommending meal plans, focus on practical implementation and flavor.`;
    
    const messages = [
      { role: 'system', content: systemPrompt },
      { role: 'user', content: userInput }
    ];
    
    const response = await this.sendPrompt(messages);
    return response.choices[0].message.content;
  }

  /**
   * Generate a meal plan based on user preferences
   * @param {Object} preferences User's dietary preferences and restrictions
   * @returns {Promise<Object>} Structured meal plan
   */
  async generateMealPlan(preferences) {
    const preferencesText = JSON.stringify(preferences);
    
    const systemPrompt = `You are a nutrition expert specializing in longevity-focused meal planning.
    Create a detailed meal plan that optimizes for longevity while respecting the user's preferences and restrictions.
    Include breakfast, lunch, dinner, and snacks for each day.
    For each meal, include:
    1. Name of the dish
    2. Brief description
    3. Key ingredients
    4. Approximate nutrition facts (calories, protein, carbs, fats)
    5. Longevity benefits

    Format the response as a structured JSON object with days and meals.`;
    
    const userPrompt = `Create a 3-day meal plan optimized for longevity based on these preferences: ${preferencesText}`;
    
    const messages = [
      { role: 'system', content: systemPrompt },
      { role: 'user', content: userPrompt }
    ];
    
    const response = await this.sendPrompt(messages);
    return response.choices[0].message.content;
  }

  /**
   * Generate a shopping list from a meal plan
   * @param {Object} mealPlan Meal plan object
   * @returns {Promise<Object>} Categorized shopping list
   */
  async generateShoppingList(mealPlan) {
    const mealPlanText = typeof mealPlan === 'object' ? JSON.stringify(mealPlan) : mealPlan;
    
    const systemPrompt = `You are a nutrition expert specializing in meal planning.
    Create a comprehensive, categorized shopping list based on the provided meal plan.
    Group items by category (e.g., produce, protein, grains, etc.).
    Include quantities where appropriate.
    Format the response as a structured JSON object.`;
    
    const userPrompt = `Create a shopping list for this meal plan: ${mealPlanText}`;
    
    const messages = [
      { role: 'system', content: systemPrompt },
      { role: 'user', content: userPrompt }
    ];
    
    const response = await this.sendPrompt(messages);
    return response.choices[0].message.content;
  }

  /**
   * Handle API errors
   * @private
   * @param {Error} error Error object
   */
  _handleError(error) {
    if (error.response) {
      // The request was made and the server responded with a status code
      // that falls out of the range of 2xx
      console.error(`API Error (${error.response.status}):`, error.response.data);
      throw new Error(`Venice API Error: ${error.response.status} - ${JSON.stringify(error.response.data)}`);
    } else if (error.request) {
      // The request was made but no response was received
      console.error('API Request Error (No response):', error.request);
      throw new Error('Venice API Request Error: No response received');
    } else {
      // Something happened in setting up the request
      console.error('API Setup Error:', error.message);
      throw new Error(`Venice API Setup Error: ${error.message}`);
    }
  }
}

export default VeniceApiClient;

File: src/api_client.py
# CulinaryAI/src/api_client.py
import os
import json
import http.client
from typing import Dict, List, Any, Optional, Union

class VeniceApiClient:
    """Client for interacting with the Venice API"""
    
    def __init__(self, api_key=None):
        """Initialize the Venice API client"""
        self.api_key = api_key or "MAT6kFr2sQebCawiHtrgy7dHgdX5yndstF8UeY7KKV"
        self.host = "api.venice.ai"
        self.model = "deepseek-r1-llama-70b"  # Default model
    
    def list_models(self) -> Dict[str, Any]:
        """
        List available models from the Venice API
        
        Returns:
            Dict: List of available models and their details
        """
        conn = http.client.HTTPSConnection(self.host)
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }
        
        try:
            conn.request("GET", "/api/v1/models", "", headers)
            response = conn.getresponse()
            data = response.read().decode("utf-8")
            
            if response.status != 200:
                self._handle_error(response.status, data)
                
            return json.loads(data)
        except Exception as e:
            raise Exception(f"Error listing models: {str(e)}")
        finally:
            conn.close()
    
    def send_prompt(
        self, 
        messages: List[Dict[str, str]], 
        options: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Send a prompt to the Venice API
        
        Args:
            messages: List of message objects with role and content
            options: Additional options for the API call
            
        Returns:
            Dict: API response
        """
        conn = http.client.HTTPSConnection(self.host)
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }
        
        default_options = {
            "temperature": 0.7,
            "max_tokens": 1000,
            "stream": False,
        }
        
        request_options = {**default_options, **(options or {})}
        
        payload = {
            "model": self.model,
            "messages": messages,
            **request_options
        }
        
        try:
            conn.request(
                "POST", 
                "/api/v1/chat/completions", 
                json.dumps(payload), 
                headers
            )
            response = conn.getresponse()
            data = response.read().decode("utf-8")
            
            if response.status != 200:
                self._handle_error(response.status, data)
                
            return json.loads(data)
        except Exception as e:
            raise Exception(f"Error sending prompt: {str(e)}")
        finally:
            conn.close()
    
    def get_nutrition_guidance(self, user_input: str) -> str:
        """
        Generate nutrition-focused response
        
        Args:
            user_input: User's question or request
            
        Returns:
            str: LLM response
        """
        system_prompt = """You are a nutrition expert specializing in longevity-focused meal planning. 
        Provide evidence-based guidance on foods and dietary patterns that promote health and longevity. 
        Emphasize whole, nutrient-dense foods like those found in Blue Zones and the Mediterranean diet.
        When recommending meal plans, focus on practical implementation and flavor."""
        
        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_input}
        ]
        
        response = self.send_prompt(messages)
        return response["choices"][0]["message"]["content"]
    
    def _handle_error(self, status_code: int, response_data: str) -> None:
        """
        Handle API errors
        
        Args:
            status_code: HTTP status code
            response_data: Response data as string
        """
        error_message = f"Venice API Error ({status_code}): {response_data}"
        print(error_message)
        raise Exception(error_message)